sdk/src/azure/platform/az_curl.c

```diff
@@ -480,7 +465,7 @@ _az_http_client_curl_setup_url(CURL* ref_curl, _az_http_request const* request)
   az_span request_url = { 0 };
   // get request_url. It will have the size of what it has written in it only
   AZ_RETURN_IF_FAILED(az_http_request_get_url(request, &request_url));
-  int32_t request_url_size = az_span_size(request_url);
+  int32_t request_url_size = _az_span_url_encode_calc_length(request_url);
```

Comment:
It's somewhat unfortunate that we have to do a double pass over the URL, once to calculate the length and again to actually encode it.

I wonder if it is worth exploring some alternatives:
1) Always assume worst case and malloc 3x unescaped url size.
2) Have the `_az_span_url_encode_calc_length` return the index when the first character to be url encoded is found, so at least we save the double processing until then.
3) Maybe we use a temporary scratch buffer and encode into that up-front, prior to the call to malloc.

Reply:
Well, we are targeting small devices,  I would rather same memory than making a double pass.